src/tiled/editpolygontool.cpp

```diff
@@ -870,3 +890,38 @@ void EditPolygonTool::splitSegments()
     if (macroStarted)
         undoStack->endMacro();
 }
+
+void EditPolygonTool::deleteEdge()
+{
+    if (mSelectedHandles.size() != 2)
+        return;
+
+    const auto &firstHandle = *mSelectedHandles.begin();
+    const auto &secondHandle = *(mSelectedHandles.begin() + 1);
+
+    const MapObjectItem *item = firstHandle->mapObjectItem();
+    MapObject *mapObject = item->mapObject();
+
+    QPolygonF polygon = mapObject->polygon();
+    QPolygonF newPolygon(polygon);
+
+    int indexDifference = std::abs(firstHandle->pointIndex() - secondHandle->pointIndex());
+
+    if (indexDifference != polygon.size() - 1) {
+        int maxIndex = std::max(firstHandle->pointIndex(), secondHandle->pointIndex());
+        for (int i = maxIndex; i < polygon.size(); ++i)
+            newPolygon[i - maxIndex] = polygon[i];
+
+        for (int i = 0; i < maxIndex; ++i)
+            newPolygon[polygon.size() - maxIndex + i] = polygon[i];
+    }
+
+    setSelectedHandles(QSet<PointHandle*>());
+
+    mapDocument()->mapObjectModel()->setObjectPolygon(mapObject, newPolygon);
```

Comment:
Instead of setting it here, you could use the `ChangePolygon` constructor that takes both the old and the new polygon.